// @dbml/core/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CompilerError", {
  enumerable: true,
  get: function get() {
    return _error.CompilerError;
  }
});
Object.defineProperty(exports, "ModelExporter", {
  enumerable: true,
  get: function get() {
    return _ModelExporter["default"];
  }
});
Object.defineProperty(exports, "Parser", {
  enumerable: true,
  get: function get() {
    return _Parser["default"];
  }
});
Object.defineProperty(exports, "VERSION", {
  enumerable: true,
  get: function get() {
    return _version.VERSION;
  }
});
Object.defineProperty(exports, "exporter", {
  enumerable: true,
  get: function get() {
    return _export["default"];
  }
});
Object.defineProperty(exports, "importer", {
  enumerable: true,
  get: function get() {
    return _import["default"];
  }
});
var _ModelExporter = _interopRequireDefault(require("./export/ModelExporter"));
var _Parser = _interopRequireDefault(require("./parse/Parser"));
var _error = require("../lib/parse/error");
var _import = _interopRequireDefault(require("./import"));
var _export = _interopRequireDefault(require("./export"));
var _version = require("./utils/version");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
